### PR TITLE
rbd: support rbd on FreeBSD (without KRBD)

### PR DESCRIPTION
**Pull request by wjwithagen**
Cmake changes to exclude KRBD and keyutils from building because FreeBSD does not support that.
Code changes to:

- Stub krbd kernel calls
- small incompatabilities, missing includes.
- Fix Clang casts errors

Signed-off-by: Willem Jan Withagen <wjw@digiware.nl>

### CODE REVIEW

**Comment by wjwithagen** [comment]
@tchaikov @liewegas 

Hi I'm getting this casting error in Clang that doesn't want to go away, with any of the casting version, other than static_cast. 
Depressing is that GNU really doesn't like that, so either we fix the code, or I if-else-endif it.

I hope you have some suggestions.

```
/usr/srcs/Ceph/work/ceph/src/test/librbd/test_mock_fixture.h:28:27: error: reinterpret_cast from 'nullptr_t' to 'ContextWQ *' is not allowed
  ContextWQ *context_wq = reinterpret_cast<ContextWQ *>(wq);
                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/srcs/Ceph/work/ceph/src/googletest/googlemock/include/gmock/gmock-generated-actions.h:523:27: note: in instantiation of function template specialization 'CompleteContextActionP2<int, nullptr_t>::gmock_Impl<void (Context *)>::gmock_PerformImpl<Context *, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg>' requested here
    return impl->template gmock_PerformImpl<A0>(args, get<0>(args),
                          ^
```

**Comment by tchaikov** [comment]
@wjwithagen i will take a look at it tomorrow.

**Comment by wjwithagen** [comment]
@tchaikov 

Probably it is due to NULL usuage on:
`/usr/srcs/Ceph/work/ceph/src/test/librbd/test_mock_Journal.cc:268:29`

But as usual I'm not able to fix it. nullptr/NULL is not quite my thing.

File: src/include/compat.h [hunk 1]
```diff
@@ -38,6 +38,9 @@
 #undef ENODATA
 #define ENODATA ENOATTR
 #endif
+#ifndef EREMOTEIO
```
Comment (tchaikov) [inline src/include/compat.h]: could you consolidate this `#define` with the ones at line:79. so there is only one place that we define `EREMOTEIO` and `HOST_NAME_MAX`. also instead of `HOST_NAME_MAX `, FreeBSD has `MAXHOSTNAMELEN`. maybe we can define `HOST_NAME_MAX` like
```c
#ifndef HOST_NAME_MAX
# ifdef MAXHOSTNAMELEN)
#  define HOST_NAME_MAX MAXHOSTNAMELEN
# else
#  define HOST_NAME_MAX 64
# endif
#endif
```
BTW, 64 is a very conservative number. probably we can make it 255.

File: src/tools/rbd/action/Kernel.cc [hunk 2]
```diff
@@ -163,6 +163,7 @@ static int parse_unmap_options(char *options)
 
 static int do_kernel_showmapped(Formatter *f)
 {
+#if defined(__linux__)
```
Comment (tchaikov) [inline src/tools/rbd/action/Kernel.cc]: probably you can just ignore `Kernel.cc` and `Nbd.cc` in `rbd_srcs` in `src/tools/rbd/CMakeLists.txt` if `!WITH_KRBD` .

File: src/test/librbd/test_mock_fixture.h [hunk 3]
```diff
@@ -25,7 +25,11 @@ ACTION_P(CopyInBufferlist, str) {
 }
 
 ACTION_P2(CompleteContext, r, wq) {
+#if defined(__clang__)
+  ContextWQ *context_wq = static_cast<ContextWQ *>(wq);
```
Comment (tchaikov) [inline src/test/librbd/test_mock_fixture.h]: although this is a little bit awkward, but this is the best we can have with minimal change, otherwise we need to change some of the callers of `CompleteContext` to cast the `NULL` parameter to the `ContextWQ *`.

@dillaman what do you think?

**Comment by wjwithagen** [comment]
@tchaikov 
Thanx for the review... I'll go over them and fix.

File: src/librbd/CMakeLists.txt [hunk 4]
```diff
@@ -107,9 +107,12 @@ target_link_libraries(librbd LINK_PRIVATE
   cls_journal_client 
   common
   pthread
-  udev
   ${CMAKE_DL_LIBS}
   ${EXTRALIBS})
+if(HAVE_UDEV)
```
Comment (wjwithagen) [inline src/librbd/CMakeLists.txt]: @tchaikov 
To start it should use the define for that:
`./cmake/modules/Findudev.cmake:23:# UDEV_LIBRARIES - List of libraries when using udev.`
But I feel uncomfortable removing it, since it was already there, and I try to change as little functionality as possible when adding FreeBSD/Clang specifics.
Old behaviour should be kept as much as possible for all others.

That said, you are right: librbd add udev.
I'll make a separate PR for that, and rebase this one.

File: src/test/cls_rbd/CMakeLists.txt [hunk 5]
```diff
@@ -14,9 +14,16 @@ target_link_libraries(ceph_test_cls_rbd
...
-  blkid
-  udev)
+  radostest)
+if(HAVE_BLKID)
+  target_link_libraries(ceph_test_cls_rbd
+    blkid)
+endif()
+if(HAVE_UDEV)
+  target_link_libraries(ceph_test_cls_rbd
+    udev)
```
Comment (wjwithagen) [inline src/test/cls_rbd/CMakeLists.txt]: @tchaikov 
Yup, like above

File: CMakeLists.txt [hunk 6]
```diff
@@ -233,6 +233,9 @@ option(WITH_KVS "Key value store is here" ON)
 # remote block storage
 option(WITH_RBD "Remote block storage is here" ON)
 
+# KERNEL remote block storage
+option(WITH_KRBD "Kernel Remote block storage is here" ON)
```
Comment (wjwithagen) [inline CMakeLists.txt]: @tchaikov 
I'll add some diganostics

File: src/tools/rbd/action/Kernel.cc (hunk 2 above)
Comment (wjwithagen) [inline src/tools/rbd/action/Kernel.cc]: @tchaikov 
I'm leaving this in for the moment, because I would like to not change the run-cli tests.
And that expects `showmapped map unmap`  in the help output.
So for met that would make this the most transparent fix.

File: src/test/librbd/test_mock_fixture.h (hunk 3 above)
Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @tchaikov @dillaman 
I'm not a real fan of this sollution either,
But it is the least invasive, and still very obvious as to why this is done.

File: src/librbd/CMakeLists.txt (hunk 4 above)
Comment (tchaikov) [inline src/librbd/CMakeLists.txt]: > I try to change as little functionality as possible when adding FreeBSD/Clang specifics.

i agree, but behaviorwise, removing libudev here does not change anything.

> I'll make a separate PR for that, and rebase this one.

thanks.

File: src/tools/rbd/action/Kernel.cc (hunk 2 above)
Comment (tchaikov) [inline src/tools/rbd/action/Kernel.cc]: makes sense. it's non-trivial to conditionalize the run-cli tests, i guess.

File: src/include/compat.h [hunk 7]
```diff
@@ -72,11 +67,18 @@
...
+/* get PATH_MAX */
 #include <limits.h>
+
+#ifndef EREMOTEIO
 #define EREMOTEIO 121
+#endif
+
+#ifndef HOST_NAME_MAX
 #define HOST_NAME_MAX 255
+#endif
```
Comment (tchaikov) [inline src/include/compat.h]: see https://github.com/ceph/ceph/pull/12798#discussion_r94904475, could use `MAXHOSTNAMELEN` instead hardwire `HOST_NAME_MAX` to 255.

Comment (tchaikov) [inline src/include/compat.h]: and could move the `#endif /* __APPLE__ */` before `#ifndef EREMOTEIO` and change it to `#endif`.

File: CMakeLists.txt (hunk 6 above)
Comment (tchaikov) [inline CMakeLists.txt]: this is not addressed.

**Comment by tchaikov** [comment]
@wjwithagen needs rebase.

**Comment by wjwithagen** [comment]
@tchaikov 
Yup, due to the UDEV commit. rebased.
I'm in and out of the datacenter with screwdriver and more, So this needs to go inbetween real work. :)

File: src/test/librbd/test_mock_fixture.h (hunk 3 above)
Comment (dillaman) [inline src/test/librbd/test_mock_fixture.h]: Hmm --- it's pretty ugly so I might vote for just using a C-style cast if Clang doesn't behave properly.

Comment (dillaman) [inline src/test/librbd/test_mock_fixture.h]: ... do you have any details on the compile failure? The only way I can see this failing is if somehow ```wq``` auto-resolved to ```nullptr_t```, but nothing that I can find is passing in ```nullptr``` (unless Clang is somehow substituting ```NULL``` for ```nullptr``` -- which is wrong since they are not the same)

File: src/tools/rbd/action/Kernel.cc (hunk 2 above)
Comment (dillaman) [inline src/tools/rbd/action/Kernel.cc]: Should it be based on ```WITH_KRBD```?

File: CMakeLists.txt [hunk 8]
```diff
@@ -232,6 +232,13 @@ option(WITH_KVS "Key value store is here" ON)
...
+
+if(WITH_KRBD AND !WITH_RBD)
+  message(FATAL_ERROR "Cannot have WITH_KRBD with WITH_RBD.")
```
Comment (tchaikov) [inline CMakeLists.txt]: s/with/without/

File: src/tools/rbd/action/Kernel.cc (hunk 2 above)
Comment (tchaikov) [inline src/tools/rbd/action/Kernel.cc]: @wjwithagen could you add `WITH_KRBD` to `config-h.in.cmake` and conditionalize on it instead?

File: CMakeLists.txt (hunk 8 above)
Comment (wjwithagen) [inline CMakeLists.txt]: @tchaikov 
Every day something new to learn.... ;-)

File: src/test/librbd/test_mock_fixture.h (hunk 3 above)
Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @dillaman @tchaikov 
This is the full output of the build error when I use the original cast:
```
 Building CXX object src/test/librbd/CMakeFiles/unittest_librbd.dir/test_mock_Journal.cc.o
cd /usr/srcs/Ceph/work/ceph/build/src/test/librbd && ccache /usr/bin/CC   -DCEPH_LIBDIR=\"/usr/local/lib\" -DCEPH_PKGLIBDIR=\"/usr/local/lib/ceph\" -DTEST_LIBRBD_INTERNALS -I/usr/srcs/Ceph/work/ceph/build/src/include -I/usr/srcs/Ceph/work/ceph/src -isystem /usr/local/include -isystem /usr/srcs/Ceph/work/ceph/build/include -I/usr/srcs/Ceph/work/ceph/src/xxHash -I/usr/srcs/Ceph/work/ceph/src/googletest/googlemock/include -I/usr/srcs/Ceph/work/ceph/src/googletest/googletest/include  -Wall -Wtype-limits -Wignored-qualifiers -Winit-self -Wpointer-arith -Werror=format-security -fno-strict-aliasing -fsigned-char -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register  -ftemplate-depth-1024 -Wno-invalid-offsetof -Wnon-virtual-dtor -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -fdiagnostics-color=auto -I/usr/local/include/nss/nss -I/usr/local/include/nspr -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wno-null-dereference -O0 -g -fPIE   -DHAVE_CONFIG_H -D__CEPH__ -D_REENTRANT -D_THREAD_SAFE -D__STDC_FORMAT_MACROS -I/usr/srcs/Ceph/work/ceph/src/googletest/googlemock/include -I/usr/srcs/Ceph/work/ceph/build/src/googletest/googlemock/include -I/usr/srcs/Ceph/work/ceph/src/googletest/googletest/include -I/usr/srcs/Ceph/work/ceph/build/src/googletest/googletest/include -fno-strict-aliasing -std=c++11 -o CMakeFiles/unittest_librbd.dir/test_mock_Journal.cc.o -c /usr/srcs/Ceph/work/ceph/src/test/librbd/test_mock_Journal.cc
In file included from /usr/srcs/Ceph/work/ceph/src/test/librbd/test_mock_Journal.cc:4:
/usr/srcs/Ceph/work/ceph/src/test/librbd/test_mock_fixture.h:31:27: error: reinterpret_cast from 'nullptr_t' to 'ContextWQ *' is not allowed
  ContextWQ *context_wq = reinterpret_cast<ContextWQ *>(wq);
                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/srcs/Ceph/work/ceph/src/googletest/googlemock/include/gmock/gmock-generated-actions.h:523:27: note: in instantiation of function template specialization 'CompleteContextActionP2<int, nullptr_t>::gmock_Impl<void (Context *)>::gmock_PerformImpl<Context *, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg, testing::internal::ExcessiveArg>' requested here
    return impl->template gmock_PerformImpl<A0>(args, get<0>(args),
                          ^
/usr/srcs/Ceph/work/ceph/src/test/librbd/test_mock_fixture.h:27:1: note: in instantiation of function template specialization 'testing::internal::ActionHelper<void, CompleteContextActionP2<int, nullptr_t>::gmock_Impl<void (Context *)> >::Perform<Context *>' requested here
ACTION_P2(CompleteContext, r, wq) {
^
/usr/srcs/Ceph/work/ceph/src/googletest/googlemock/include/gmock/gmock-generated-actions.h:1420:13: note: expanded from macro 'ACTION_P2'
            Perform(this, args);\
            ^
/usr/srcs/Ceph/work/ceph/src/test/librbd/test_mock_fixture.h:27:1: note: in instantiation of member function 'CompleteContextActionP2<int, nullptr_t>::gmock_Impl<void (Context *)>::Perform' requested here
/usr/srcs/Ceph/work/ceph/src/googletest/googlemock/include/gmock/gmock-generated-actions.h:1416:7: note: expanded from macro 'ACTION_P2'
      gmock_Impl(p0##_type gmock_p0, p1##_type gmock_p1) : p0(gmock_p0), \
      ^
/usr/srcs/Ceph/work/ceph/src/test/librbd/test_mock_fixture.h:27:1: note: in instantiation of member function 'CompleteContextActionP2<int, nullptr_t>::gmock_Impl<void (Context *)>::gmock_Impl' requested here
/usr/srcs/Ceph/work/ceph/src/googletest/googlemock/include/gmock/gmock-generated-actions.h:1436:39: note: expanded from macro 'ACTION_P2'
      return ::testing::Action<F>(new gmock_Impl<F>(p0, p1));\
                                      ^
/usr/srcs/Ceph/work/ceph/src/test/librbd/test_mock_Journal.cc:268:29: note: in instantiation of function template specialization 'CompleteContextActionP2<int, nullptr_t>::operator Action<void (Context *)>' requested here
                  .WillOnce(CompleteContext(0, NULL));
                            ^
1 error generated.
```

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @dillaman @tchaikov 
And I think it is because compiletime evaluation of:
`WillOnce(CompleteContext(0, NULL)`
Delivers a nullptr, and that is where the trouble starts?

**Comment by wjwithagen** [comment]
@tchaikov 
I think I addressed your suggestions...

File: src/test/librbd/test_mock_fixture.h (hunk 3 above)
Comment (tchaikov) [inline src/test/librbd/test_mock_fixture.h]: weird enough, with clang++-4.0, clang translates translate`NULL` to `__null` with
```
clang++ -std=c++11 -E -o test.C test.cc
```

with following sample c++ source file
```c++
#include <cstddef>

NULL;
```

please note that `libstd++` defines `NULL` to `__null` in `stddef.h`.

quote from `stddef.h` from llvm.
```c
#if defined(__need_NULL)
#undef NULL
#ifdef __cplusplus
#  if !defined(__MINGW32__) && !defined(_MSC_VER)
#    define NULL __null
#  else
#    define NULL 0
#  endif
#else
#  define NULL ((void*)0)
#endif
```
see https://github.com/llvm-mirror/clang/blob/master/lib/Headers/stddef.h#L96.

@wjwithagen could you double check your building environment?

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @tchaikov 
You might be onto something, last set of lines:
```
# 44 "/usr/include/c++/v1/cstddef" 2 3
# 1 "/usr/include/c++/v1/__nullptr" 1 3
# 18 "/usr/include/c++/v1/__nullptr" 3
# 59 "/usr/include/c++/v1/__nullptr" 3
namespace std
{
    typedef decltype(nullptr) nullptr_t;
}
....
# 2 "test-NULL.cc" 2
nullptr;
```

But I think Clang uses a different C++ lib:
[https://wiki.freebsd.org/NewC++Stack](https://wiki.freebsd.org/NewC++Stack)

Which is were the difference originates ... 
How you/we would like to solve this in Ceph is a question...

I would still vote to use the #if-else-end

Comment (tchaikov) [inline src/test/librbd/test_mock_fixture.h]: @wjwithagen please note, i was indeed checking the LLVM's libc++ instead of the GNU's libstdc++ . maybe you could check your `stddef.h` to see how `NULL` is defined?

Comment (tchaikov) [inline src/test/librbd/test_mock_fixture.h]: if the libc++'s `stddef.h` looks sane. maybe we should pass to `-stdlib=libc++` the compiler on FreeBSD to use the "new" c++11 stack?

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @tchaikov 
Ah, I see, will check.

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @tchaikov

I'm expecting clang on FreeBSD to use the new libc++ stack per default.
Als look at the current include order:
```
> clang++ -v -std=c++11 -E -o test-NULL.C test-NULL.cc
FreeBSD clang version 3.8.0 (tags/RELEASE_380/final 262564) (based on LLVM 3.8.0)
Target: x86_64-unknown-freebsd12.0
Thread model: posix
InstalledDir: /usr/bin
 "/usr/bin/clang++" -cc1 -triple x86_64-unknown-freebsd12.0 -E -disable-free -main-file-name test-NULL.cc -mrelocation-model static -mthread-model posix -mdisable-fp-elim -masm-verbose -mconstructor-aliases -munwind-tables -target-cpu x86-64 -v -dwarf-column-info -debugger-tuning=gdb -resource-dir /usr/bin/../lib/clang/3.8.0 -internal-isystem /usr/include/c++/v1 -std=c++11 -fdeprecated-macro -fdebug-compilation-dir /home/wjw/Src/c-code/div -ferror-limit 19 -fmessage-length 132 -fobjc-runtime=gnustep -fcxx-exceptions -fexceptions -fdiagnostics-show-option -fcolor-diagnostics -o test-NULL.C -x c++ test-NULL.cc
clang -cc1 version 3.8.0 based upon LLVM 3.8.0 default target x86_64-unknown-freebsd12.0
#include "..." search starts here:
#include <...> search starts here:
 /usr/include/c++/v1
 /usr/bin/../lib/clang/3.8.0/include
 /usr/include
End of search list.
```
Which is the same with or without `-stdlib=libc++` 

But even following the `/usr/include` files:

`/usr/include/stddef.h` -> `sys/_null.h`
`sys/_null.h`:
```
#if !defined(__cplusplus)
#define NULL    ((void *)0)
#else
#if __cplusplus >= 201103L
#define NULL    nullptr
#elif defined(__GNUG__) && defined(__GNUC__) && __GNUC__ >= 4
#define NULL    __null
#else
#if defined(__LP64__)
#define NULL    (0L)
#else
#define NULL    0
#endif  /* __LP64__ */
#endif  /* __GNUG__ */
#endif  /* !__cplusplus */
```
Which would hit the `#if __cplusplus >= 201103L` line, and we end up (again) with:
``#define     NULL    nullptr`

Comment (tchaikov) [inline src/test/librbd/test_mock_fixture.h]: @wjwithagen seems on FreeBSD the C land is still using the FreeBSD's own libc. see 

- stdef.h from llvm/libc++: https://github.com/llvm-mirror/clang/blob/master/lib/Headers/stddef.h and 
- stddef.h from FreeBSD. https://github.com/freebsd/freebsd/blob/386ddae58459341ec567604707805814a2128a57/sys/sys/stddef.h
  - and included _null.h: https://github.com/freebsd/freebsd/blob/386ddae58459341ec567604707805814a2128a57/sys/sys/_null.h

in other words, FreeBSD folks think that `NULL == nullptr` with C++11 or up. per N3337.

§ 18.2
> The macro NULL is an implementation-defined C++ null pointer constant in this International Standard (4.10).194

§ 4.10
> 4.10 Pointer conversions [conv.ptr]
> A null pointer constant is an integral constant expression (5.19) prvalue of integer type that evaluates to zero **or** a prvalue of type std::nullptr_t. A null pointer constant can be converted to a pointer type; the result is the null pointer value of that type and is distinguishable from every other value of object pointer or function pointer type. Such a conversion is called a null pointer conversion. Two null pointer values of the same type shall compare equal. The conversion of a null pointer constant to a pointer to cv-qualified type is a single conversion, and not the sequence of a pointer conversion followed by a qualification conversion (4.4).
> A null pointer constant of integral type can be converted to a prvalue of type std::nullptr_t. [ Note: The resulting prvalue is not a null pointer value. — end note ]

so i guess that FreeBSD's implementation of NULL makes sense. as it choose to use "a prvalue of type std::nullptr_t" instead of "an integral constant expression (5.19) prvalue of integer type that evaluates to zero". and the former is also acceptable per the C++11 standard.

@dillaman does this make sense?

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @tchaikov 

The result will stay the same, but for userspace includes you'd want to look at:
[https://github.com/freebsd/freebsd/blob/386ddae58459341ec567604707805814a2128a57/include/stddef.h](https://github.com/freebsd/freebsd/blob/386ddae58459341ec567604707805814a2128a57/include/stddef.h).
Because sys/sys are the kernel include files, and they can differ.

Comment (tchaikov) [inline src/test/librbd/test_mock_fixture.h]: @wjwithagen thanks, fixed my comment.

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @tchaikov @dillaman 
Any suggestions on how to move this forward?

1) Rip open all the other calls and start using nullptr everywhere?
2) Let this part be ifdef-ed 
2a) Let this part be ifdef-ed for the time being, and work on 1)

Comment (tchaikov) [inline src/test/librbd/test_mock_fixture.h]: @dillaman IIUC, you suggested to use the C style cast, right?

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @tchaikov 
Willing to test, but I rather clueless as to exactly what and where, whithout upsetting even more code. AND is also not rejected by either GNU of Clang.

Comment (dillaman) [inline src/test/librbd/test_mock_fixture.h]: I think you will get away w/ just using a ```static_cast``` for both cases

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @dillaman 
Are you suggesting to `static_cast`: `  .WillOnce(CompleteContext(0, NULL));` ??

Comment (dillaman) [inline src/test/librbd/test_mock_fixture.h]: No -- I am suggesting to keep your new ```ContextWQ *context_wq = static_cast<ContextWQ *>(wq)``` line for all cases and eliminate the ```#ifdef```

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @dillaman 
When I did that then GNUCC started tripping...
Have to start another run in my CentOS box for that.

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @tchaikov @dillaman 
Clang is happy when I keep the reinterpret_cast for `ContextWQ *context_wq = static_cast<ContextWQ *>(wq)` 
And then fix `CompleteContext(0, NULL)` => `CompleteContext(0, static_cast<ContextWQ *>(NULL))
But I'm not yet sure that GNUCC is going to like that.

Comment (dillaman) [inline src/test/librbd/test_mock_fixture.h]: @wjwithagen In that specific case above, the ```reinterpret_cast``` is a no-op since the argument already resolves to a ```ContextWQ *```

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @dillaman 
My Linux VM is not really fast, so compiling will take some time.

I would expect so, but when I changed it in the template, GNUCC was not happy.
I'll get back here with GNUCC's complaints. (or not)

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @dillaman 
Compiling GnuCC gives:
```
In file included from /home/wjw/wip/src/test/librbd/test_mock_Journal.cc:4:0:
/home/wjw/wip/src/test/librbd/test_mock_fixture.h: In instantiation of ‘typename testing::internal::Function<F>::Result CompleteContextActionP2<r_type, wq_type>::gmock_Impl<F>::gmock_PerformImpl(const args_type&, arg0_type, arg1_type, arg2_type, arg3_type, arg4_type, arg5_type, arg6_type, arg7_type, arg8_type, arg9_type) const [with arg0_type = Context*; arg1_type = testing::internal::ExcessiveArg; arg2_type = testing::internal::ExcessiveArg; arg3_type = testing::internal::ExcessiveArg; arg4_type = testing::internal::ExcessiveArg; arg5_type = testing::internal::ExcessiveArg; arg6_type = testing::internal::ExcessiveArg; arg7_type = testing::internal::ExcessiveArg; arg8_type = testing::internal::ExcessiveArg; arg9_type = testing::internal::ExcessiveArg; F = void(Context*); r_type = int; wq_type = long int; typename testing::internal::Function<F>::Result = void; CompleteContextActionP2<r_type, wq_type>::gmock_Impl<F>::args_type = std::tuple<Context*>]’:
/home/wjw/wip/src/googletest/googlemock/include/gmock/gmock-generated-actions.h:526:23:   required from ‘static Result testing::internal::ActionHelper<Result, Impl>::Perform(Impl*, const std::tuple<A0>&) [with A0 = Context*; Result = void; Impl = CompleteContextActionP2<int, long int>::gmock_Impl<void(Context*)>]’
/home/wjw/wip/src/test/librbd/test_mock_fixture.h:27:1:   required from ‘CompleteContextActionP2<r_type, wq_type>::gmock_Impl<F>::return_type CompleteContextActionP2<r_type, wq_type>::gmock_Impl<F>::Perform(const args_type&) [with F = void(Context*); r_type = int; wq_type = long int; CompleteContextActionP2<r_type, wq_type>::gmock_Impl<F>::return_type = void; CompleteContextActionP2<r_type, wq_type>::gmock_Impl<F>::args_type = std::tuple<Context*>]’
/home/wjw/wip/src/test/librbd/test_mock_Journal.cc:1349:1:   required from here
/home/wjw/wip/src/test/librbd/test_mock_fixture.h:32:54: error: invalid static_cast from type ‘const long int’ to type ‘ContextWQ*’
   ContextWQ *context_wq = static_cast<ContextWQ *>(wq);
```

Which is probably again due to the way that in linux NULL is defined...

Comment (dillaman) [inline src/test/librbd/test_mock_fixture.h]: Well that's annoying -- ugly #ifdef it is for now I guess

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @dillaman 
Not quite.....
A nicer sollution might be to actually fix the typecaset for `NULL`:
    `CompleteContext(r,  static_cast<ContextWQ *>(NULL))`

That works for both FreeBSD/Clang and linux/GCC.
Need to fix that in about 8 locations.
Could do that in a separate PR and drop this commit from this PR?

Comment (dillaman) [inline src/test/librbd/test_mock_fixture.h]: That's fine as well -- long term would still need a Clang compiler in the Jenkins build set to ensure future PRs don't break things

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @dillaman 
And it needs to be running with a environment where NULL is defined to be nullptr !!
Will submit a PR tomorrow. It is past midnight here.

Comment (tchaikov) [inline src/test/librbd/test_mock_fixture.h]: yeah, that's also what i suggested in the beginning. as NULL is could also be defined as `0` where `static_cast<>()` does not work.

Comment (wjwithagen) [inline src/test/librbd/test_mock_fixture.h]: @tchaikov @dillaman 
Gentleman,
If you could checkout #12880.
Once that is accepted, I'll rebase this PR to macht it.

**Comment by wjwithagen** [comment]
@tchaikov @dillaman 
Rebased for the NULL-cast if-def PR.
So this PR is now more or less how it should be as well.
